Run KNOWNBUG and THOROUGH regression tests in CI

tautschnig opened the pull request:
Use the check-ubuntu-20_04-cmake-gcc job to run KNOWNBUG (any test
reported as failure will tell us that a bug has unexpectedly been fixed)
an THOROUGH (tests that are expected to pass, but take longer to do so)
tests.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Run KNOWNBUG and THOROUGH regression tests in CI

Copy the check-ubuntu-20_04-cmake-gcc job to run KNOWNBUG (any test
reported as failure will tell us that a bug has unexpectedly been fixed)
an THOROUGH (tests that are expected to pass, but take longer to do so)
tests, each as a separate GitHub action.

Also run all tests tagged broken-smt-backend to confirm they haven't
been fixed.

martin-cs reviewed (approved):
This is a nice idea.  Is it worth running this once first and PRing any of the KNOWNBUG tests that now work?  My impression is there are probably going to be some and they may need attention to figure out if they are really working on just incorrect tests.

NlightNFotis commented:
I'm not a huge fan of repurposing existing jobs to be honest.

Would it be possible that this gets done as a separate job?

Otherwise I agree that this is an excellent idea.

tautschnig commented:
> I'm not a huge fan of repurposing existing jobs to be honest.
> 
> Would it be possible that this gets done as a separate job?
> 
> Otherwise I agree that this is an excellent idea.

I guess I'm still living the times of limited Travis runners :-)  That's fixed now, added two new jobs.

tautschnig commented:
> This is a nice idea. Is it worth running this once first and PRing any of the KNOWNBUG tests that now work? My impression is there are probably going to be some and they may need attention to figure out if they are really working on just incorrect tests.

Belated response: yes, I'll try to `git bisect` to figure out when exactly tests were fixed. Will keep this PR in "Draft" state for the time being.

tautschnig committed:
Mark failing goto-instrument-wmm tests as KNOWNBUG

These had been marked as THOROUGH without fully checking their status.

tautschnig committed:
Fix line number to make location15 test pass as expected

The block includes lines 10 and 11.

tautschnig committed:
jbmc THOROUGH regression tests: fix spec files

Several tests spuriously failed as they were lacking the Java models
library. Some of them also do not need to be marked as "THOROUGH" for
they terminate in under 1 second. One test moved from THOROUGH to
KNOWNBUG as the assertions are (unexpectedly) deemed unreachable.

tautschnig committed:
gcc_popcount2: fix assertion matching pattern

The assertion of interest now has number 3, which is irrelevant and
should not cause the test to fail.

tautschnig committed:
jbmc-strings/VerifStringLastIndexOf: ensure index is within String

The reference implementation and the built-in function disagree in case
the index is out of bounds.

TGWDB reviewed (commented):
I like the idea of the KNOWNBUG checking to routinely check for improvements/changes. The THOROUGH on all PRs seems like it may be a lot of extra load?

TGWDB commented on .github/workflows/pull-request-checks.yaml:
```diff
@@ -170,6 +170,90 @@ jobs:
+          DEBIAN_FRONTEND: noninteractive
+        run: |
+          sudo apt-get update
+          sudo apt-get install --no-install-recommends -yq cmake ninja-build gcc g++ maven flex bison libxml2-utils ccache z3
```
The new z3 output on version is not in the new check.

tautschnig replied:
Done!

TGWDB commented on .github/workflows/pull-request-checks.yaml:
```diff
@@ -170,6 +170,90 @@ jobs:
+          DEBIAN_FRONTEND: noninteractive
+        run: |
+          sudo apt-get update
+          sudo apt-get install --no-install-recommends -yq cmake ninja-build gcc g++ maven flex bison libxml2-utils ccache z3
```
z3 version output not here again

tautschnig replied:
Done!

tautschnig commented:
> I like the idea of the KNOWNBUG checking to routinely check for improvements/changes. The THOROUGH on all PRs seems like it may be a lot of extra load?

I'm not sure we care? "CBMC / codecov-coverage-report" takes longer (102 vs 75 minutes), and GitHub happily runs all of the in parallel (up to 20 - according to https://docs.github.com/en/actions/reference/usage-limits-billing-and-administration#usage-limits).

TGWDB reviewed (commented):
Broadly this PR looks good, but I would prefer that other fixes/changes that are not simply a CI change be a separate PR. For example https://github.com/diffblue/cbmc/pull/5958/commits/00847391c31f6e3fd2781eb379fa244656084420 appears to be bug fix that was found along the way (and great to have done), but not part of the titular change.

tautschnig commented:
> Broadly this PR looks good, but I would prefer that other fixes/changes that are not simply a CI change be a separate PR. For example [0084739](https://github.com/diffblue/cbmc/commit/00847391c31f6e3fd2781eb379fa244656084420) appears to be bug fix that was found along the way (and great to have done), but not part of the titular change.

Fair point, I've factored those out into #6102, #6103, #6104. The remaining commits could perhaps be factored out as well, but wouldn't be subject to CI-based testing until last commit of this PR is merged (which, in turn, can only be merged if fixing those issues first).

tautschnig commented:
@TGWDB All dependencies are now merged and this PR is rebased and should be ready for review.

NlightNFotis reviewed (approved):
I'm happy with this personally, provided we don't mark the extra CI jobs as required - they shouldn't be blocking PRs in my opinion.
